pytorch_lightning/trainer/lightning_config.py

```diff
@@ -0,0 +1,247 @@
+from typing import Union, Optional, List, Dict, Iterable
+from dataclasses import dataclass, fields
+
+from pytorch_lightning.loggers import LightningLoggerBase
+from pytorch_lightning.callbacks import ModelCheckpoint, EarlyStopping, Callback
+from pytorch_lightning.profiler import BaseProfiler
+
+
+@dataclass
+class LightningConfig:
```

Comment:
this may have a disadvantage that all the arguments won't be listed in Trainer docs page and so a user needs to search for it also it drops IDE support of whispering possible arguments

Reply:
I like the idea of having a lightning config. I think we could maybe change it that the trainer still has all the arguments, internally wrap it to a config holding all the internal state variables as well as configuration and just make the trainer have properties for the most relevant attributes. 

This would also have the advantage, that you could possibly save and resume the whole trainer state at every moment

Reply:
i also like the config... much easier to keep track of args and share among classes.
i do prefer allowing the init to be:
```
Trainer(arg1, arg2, etc...)
```

Instead of
```
Trainer(config)
```

Internally we should wrap it in the config object. Now, in theory people could use the config directly if they wanted to:
```
Trainer(**config)
```